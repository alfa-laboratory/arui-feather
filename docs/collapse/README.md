# Collapse

Компонент "подката": позволяет спрятать кусок текста за ссылку "Еще...".

```javascript
import Collapse from 'arui-feather/collapse';
```




## Props


| Prop  | Тип  | По-умолчанию | Обязательный | Описание |
| ----- | ---- | ------------ | ------------ |----------|
| isExpanded | Boolean |  |  | Управление состоянием `expand`/`collapse` компонента |
| collapsedLabel | String | `'Expand'`  |  | Текст ссылки в `expand` состоянии |
| expandedLabel | String | `'Collapse'`  |  | Текст ссылки в `collapse` состоянии |
| children | Array.<Node>\|Node |  |  | Дочерние элементы `Collapse` |
| theme | [ThemeEnum](#ThemeEnum) |  |  | Тема компонента |
| className | Function\|String |  |  | Дополнительный класс |
| onExpandedChange | Function |  |  | Обработчик смены состояния `expand`/`collapse` |







## Типы






### <a id="ThemeEnum"></a>ThemeEnum

 * `'alfa-on-color'`
 * `'alfa-on-white'`



