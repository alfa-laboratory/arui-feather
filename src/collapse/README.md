
```
const LOREM_IPSUM = require('../vars').LOREM_IPSUM;
<Collapse
    collapsedLabel='Подробнее'
    expandedLabel='Скрыть' >
    <span>
        { LOREM_IPSUM.slice(0, 3) }
    </span>
</Collapse>
```

Открытый
```
const LOREM_IPSUM = require('../vars').LOREM_IPSUM;
function handleExpandedChange(isExpanded){
   setState({ isExpanded });
}
initialState = {
    isExpanded: true
};
<Collapse
    collapsedLabel='Подробнее'
    expandedLabel='Скрыть'
    isExpanded={ state.isExpanded }
    onExpandedChange={ value => handleExpandedChange(value) }>
    <span>
        { LOREM_IPSUM.slice(0, 3) }
    </span>
</Collapse>
```
