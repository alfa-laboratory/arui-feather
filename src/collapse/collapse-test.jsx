/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { render, cleanUp } from '../test-utils';

import Collapse from './collapse';

const TEXT = 'Collapsing text';

describe('collapse', () => {
    afterEach(cleanUp);

    it('should render without problem', () => {
        let collapse = render(<Collapse>{ TEXT }</Collapse>);

        expect(collapse.node).to.exist;
        expect(collapse.node).to.have.class('collapse');
    });

    it('should call `onExpandedChange` callback after expand/collapse', () => {
        let onExpandedChange = sinon.spy();
        let collapse = render(
            <Collapse
                collapsedLabel='Expand'
                expandedLabel='Collapse'
                onExpandedChange={ onExpandedChange }
            >
                { TEXT }
            </Collapse>
        );
        let linkNode = collapse.node.querySelector('.link');

        linkNode.click();

        expect(onExpandedChange).to.have.been.calledOnce;
    });

    it('should update component height', (done) => {
        const HEIGHT_DELTA = 100;
        let collapse = render(
            <Collapse
                collapsedLabel='Expand'
                expandedLabel='Collapse'
            >
                <div
                    id='pusher'
                    style={ { width: '100%', height: '10px' } }
                />
            </Collapse>
        );
        let collapseContent = collapse.node.querySelector('.collapse__content');
        let linkNode = collapse.node.querySelector('.link');

        setTimeout(() => {
            document.getElementById('pusher').style.height = `${HEIGHT_DELTA}px`;

            linkNode.click();

            setTimeout(() => {
                expect(collapseContent.style.height).to.equal(`${HEIGHT_DELTA}px`);
                done();
            }, 0);
        }, 0);
    });

    it('should not update component height when it is collapsed', (done) => {
        const HEIGHT_DELTA = 100;
        let collapse = render(
            <Collapse
                collapsedLabel='Expand'
                expandedLabel='Collapse'
            >
                <div
                    id='pusher'
                    style={ { width: '100%', height: '0px' } }
                />
            </Collapse>
        );
        let initialComponentHeight = collapse.node.offsetHeight;

        document.getElementById('pusher').style.height = `${HEIGHT_DELTA}px`;

        setTimeout(() => {
            expect(collapse.node.offsetHeight).to.equal(initialComponentHeight);
            done();
        }, 1000);
    });

    it('should apply custom expanded label', () => {
        let collapse = render(
            <Collapse
                collapsedLabel='Раскрыть'
                isExpanded={ false }
            >
                { TEXT }
            </Collapse>
        );
        let linkNode = collapse.node.querySelector('.link');

        expect(linkNode).to.have.text('Раскрыть');
    });

    it('should apply custom collapsed label', () => {
        let collapse = render(
            <Collapse
                expandedLabel='Закрыть'
                isExpanded={ true }
            >
                { TEXT }
            </Collapse>
        );
        let linkNode = collapse.node.querySelector('.link');

        expect(linkNode).to.have.text('Закрыть');
    });

    it('should have default expanded label', () => {
        let collapse = render(<Collapse isExpanded={ false }>{ TEXT }</Collapse>);
        let linkNode = collapse.node.querySelector('.link');

        expect(linkNode).to.have.text('Expand');
    });

    it('should have default collapsed label', () => {
        let collapse = render(<Collapse isExpanded={ true }>{ TEXT }</Collapse>);
        let linkNode = collapse.node.querySelector('.link');

        expect(linkNode).to.have.text('Collapse');
    });
});
