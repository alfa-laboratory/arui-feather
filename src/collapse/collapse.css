/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import '../vars.css';

.collapse {
    font-family: var(--font);
    font-weight: var(--font-weight-normal);

    &, *, *:before, *:after {
        box-sizing: border-box;
    }

    &__content {
        position: relative;
    }

    &__link {
        outline: none;

        &:after {
            content: '';
            display: inline-block;
            width: 10px;
            height: 10px;
            margin-left: 5px;
            background-size: 100% 100%;
        }
    }

    &__content {
        overflow: hidden;
        visibility: hidden;
        height: 0;
        transition: height 0.2s;
    }

    &_expanded &__content {
        visibility: visible;
        height: auto;
        margin-bottom: 15px;
    }

    &_expanded &__link:after {
        transform: scaleY(-1);
    }

    &_theme_alfa-on-color &__link:after {
        background-image: url('./images/collapse__link_theme_alfa-on-color.svg');
    }

    &_theme_alfa-on-white &__link:after {
        background-image: url('./images/collapse__link_theme_alfa-on-white.svg');
    }
}
