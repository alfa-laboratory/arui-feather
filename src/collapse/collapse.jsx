/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import Type from 'prop-types';

import IconArrowDown from '../icon/ui/arrow-down';
import IconArrowUp from '../icon/ui/arrow-up';
import Link from '../link/link';
import ResizeSensor from '../resize-sensor/resize-sensor';

import cn from '../cn';
import performance from '../performance';

/**
 * Компонент «подката» позволяет спрятать кусок текста за ссылку «Еще...».
 */
@cn('collapse')
@performance()
class Collapse extends React.Component {
    static propTypes = {
        /** Управление `expanded` состоянием компонента */
        isExpanded: Type.bool,
        /** Текст ссылки в `expanded` состоянии */
        collapsedLabel: Type.string,
        /** Текст ссылки в `collapsed` состоянии */
        expandedLabel: Type.string,
        /** Дочерние элементы `Collapse` */
        children: Type.oneOfType([Type.arrayOf(Type.node), Type.node]),
        /** Тема компонента */
        theme: Type.oneOf(['alfa-on-color', 'alfa-on-white']),
        /** Дополнительный класс */
        className: Type.string,
        /** Идентификатор компонента в DOM */
        id: Type.string,
        /**
         * Обработчик смены состояний `expanded/collapsed`
         * @param {boolean} isExpanded
         */
        onExpandedChange: Type.func,
        /** Идентификатор для систем автоматизированного тестирования */
        'data-test-id': Type.string
    };

    static defaultProps = {
        expandedLabel: 'Collapse',
        collapsedLabel: 'Expand'
    };

    state = {
        isExpanded: false
    };

    content;
    contentCase;

    componentDidMount() {
        this.updateContentHeight();
    }

    componentDidUpdate() {
        this.updateContentHeight();
    }

    render(cn) {
        let ToggledIcon;
        const expanded = this.props.isExpanded === undefined ? this.state.isExpanded : this.props.isExpanded;

        switch (expanded) {
            case true: ToggledIcon = IconArrowUp; break;
            case false: ToggledIcon = IconArrowDown; break;
        }

        return (
            <div
                className={ cn({
                    expanded
                }) }
                id={ this.props.id }
                data-test-id={ this.props['data-test-id'] }
            >
                <div
                    ref={ (content) => {
                        this.content = content;
                    } }
                    className={ cn('content') }
                >
                    <div ref={ (contentCase) => {
                        this.contentCase = contentCase;
                    } }
                    >
                        { this.props.children }
                    </div>
                    <ResizeSensor onResize={ this.updateContentHeight } />
                </div>
                <Link
                    className={ cn('link') }
                    view='dark'
                    pseudo={ true }
                    icon={
                        <ToggledIcon size='s' />
                    }
                    iconPosition='right'
                    onClick={ this.handleExpandedChange }
                    text={
                        expanded
                            ? this.props.expandedLabel
                            : this.props.collapsedLabel
                    }
                />
            </div>
        );
    }

    handleExpandedChange = () => {
        const newExpandedValue = this.props.isExpanded === undefined ? !this.state.isExpanded : !this.props.isExpanded;

        this.setState({
            isExpanded: newExpandedValue
        });

        if (this.props.onExpandedChange) {
            this.props.onExpandedChange(newExpandedValue);
        }
    };

    updateContentHeight = () => {
        const expanded = this.props.isExpanded === undefined ? this.state.isExpanded : this.props.isExpanded;

        let contentHeight;

        if (expanded) {
            contentHeight = this.contentCase.offsetHeight;
        } else {
            contentHeight = 0;
        }

        if (this.content) {
            this.content.style.height = `${contentHeight}px`;
        }
    }
}

export default Collapse;
