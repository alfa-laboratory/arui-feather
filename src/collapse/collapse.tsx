/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { createCn } from 'bem-react-classname';
import { withTheme } from '../cn';

import IconArrowDown from '../icon/ui/arrow-down';
import IconArrowUp from '../icon/ui/arrow-up';
import Link from '../link/link';
import { ResizeSensor } from '../resize-sensor/resize-sensor';

export type CollapseProps = {
    /**
     * Управление `expanded` состоянием компонента
     * */
    isExpanded?: boolean;

    /**
     * Текст ссылки в `expanded` состоянии
     * */
    collapsedLabel?: string;

    /**
     * Текст ссылки в `collapsed` состоянии
     * */
    expandedLabel?: string;

    /**
     * Дочерние элементы `Collapse`
     */
    children?: ReadonlyArray<React.ReactNode> | React.ReactNode;

    /**
     * Тема компонента
     */
    theme?: 'alfa-on-color' | 'alfa-on-white';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Идентификатор компонента в DOM
     */
    id?: string;

    /**
     * Обработчик смены состояний `expanded/collapsed`
     */
    onExpandedChange?: (isExpanded?: boolean) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    'data-test-id'?: string;
};

/**
 * Компонент «подката» позволяет спрятать кусок текста за ссылку «Еще...».
 */
export class Collapse extends React.PureComponent<CollapseProps> {
    protected cn = createCn('collapse');

    static defaultProps: Partial<CollapseProps> = {
        expandedLabel: 'Collapse',
        collapsedLabel: 'Expand'
    };

    state = {
        isExpanded: false
    };

    // TODO [issues/1018] переписать тесты нужно, что бы private был
    content;
    contentCase;

    componentDidMount() {
        this.updateContentHeight();
    }

    componentDidUpdate() {
        this.updateContentHeight();
    }

    render() {
        let ToggledIcon;
        const expanded = this.props.isExpanded === undefined ? this.state.isExpanded : this.props.isExpanded;

        switch (expanded) {
            case true: ToggledIcon = IconArrowUp; break;
            case false: ToggledIcon = IconArrowDown; break;
        }

        return (
            <div
                className={ this.cn({
                    expanded
                }) }
                id={ this.props.id }
                data-test-id={ this.props['data-test-id'] }
            >
                <div
                    ref={ (content) => {
                        this.content = content;
                    } }
                    className={ this.cn('content') }
                >
                    <div ref={ (contentCase) => {
                        this.contentCase = contentCase;
                    } }
                    >
                        { this.props.children }
                    </div>
                    <ResizeSensor onResize={ this.updateContentHeight } />
                </div>
                <Link
                    className={ this.cn('link') }
                    pseudo={ true }
                    icon={
                        <ToggledIcon size='s' />
                    }
                    iconPosition='right'
                    onClick={ this.handleExpandedChange }
                    text={
                        expanded
                            ? this.props.expandedLabel
                            : this.props.collapsedLabel
                    }
                />
            </div>
        );
    }

    private handleExpandedChange = () => {
        const newExpandedValue = this.props.isExpanded === undefined ? !this.state.isExpanded : !this.props.isExpanded;

        this.setState({
            isExpanded: newExpandedValue
        });

        if (this.props.onExpandedChange) {
            this.props.onExpandedChange(newExpandedValue);
        }
    };

    private updateContentHeight = () => {
        const expanded = this.props.isExpanded === undefined ? this.state.isExpanded : this.props.isExpanded;

        let contentHeight;

        if (expanded) {
            contentHeight = this.contentCase.offsetHeight;
        } else {
            contentHeight = 0;
        }

        if (this.content) {
            this.content.style.height = `${contentHeight}px`;
        }
    }
}

export default withTheme(Collapse);
