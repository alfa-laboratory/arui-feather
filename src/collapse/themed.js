import { withTheme } from '../cn';
import Component from './collapse';

export default withTheme(Component);
