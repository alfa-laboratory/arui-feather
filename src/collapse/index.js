import '../link';
import '../resize-sensor';

import './collapse.css';

export default from './collapse';
