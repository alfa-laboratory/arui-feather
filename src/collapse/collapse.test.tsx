/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { shallow, mount } from 'enzyme';

import { Collapse } from './collapse';

describe('collapse', () => {
    it('should render without problem', () => {
        const collapse = shallow(<Collapse>Collapsing text</Collapse>);

        expect(collapse).toMatchSnapshot();
        expect(collapse.prop('className')).toContain('collapse');
    });

    it('should call `onExpandedChange` callback after expand/collapse', () => {
        const onExpandedChange = jest.fn();
        const collapse = mount(
            <Collapse
                collapsedLabel='Expand'
                expandedLabel='Collapse'
                onExpandedChange={ onExpandedChange }
            >
                Collapsing text
            </Collapse>
        );
        const linkNode = collapse.find('.link');

        linkNode.simulate('click');

        expect(onExpandedChange).toHaveBeenCalled();
    });

    it('should update component height', () => {
        const collapse = mount<Collapse>(
            <Collapse>
                some children
            </Collapse>
        );

        collapse.setState({ isExpanded: true });
        // jsdom не умеет обсчитывать высоты элементов правильным образом, поэтому
        // эмулируем высоту контент-враппера и самостоятельно вызываем resize-handler
        collapse.instance().contentCase = { offsetHeight: 100 };
        // @ts-ignore
        collapse.instance().updateContentHeight();

        expect(collapse.instance().content.style.height).toBe('100px');
    });

    it('should not update component height when it is collapsed', () => {
        const collapse = mount<Collapse>(
            <Collapse>
                some children
            </Collapse>
        );

        collapse.setState({ isExpanded: false });
        // jsdom не умеет обсчитывать высоты элементов правильным образом, поэтому
        // эмулируем высоту контент-враппера и самостоятельно вызываем resize-handler
        collapse.instance().contentCase = { offsetHeight: 100 };
        // @ts-ignore
        collapse.instance().updateContentHeight();

        expect(collapse.instance().content.style.height).toBe('0px');
    });

    it('should apply custom expanded label', () => {
        const collapse = mount(
            <Collapse
                collapsedLabel='Раскрыть'
                isExpanded={ false }
            >
                Collapsing text
            </Collapse>
        );
        const linkNode = collapse.find('.link');

        expect(linkNode.text()).toBe('Раскрыть');
    });

    it('should apply custom collapsed label', () => {
        const collapse = mount(
            <Collapse
                expandedLabel='Закрыть'
                isExpanded={ true }
            >
                Collapsing text
            </Collapse>
        );
        const linkNode = collapse.find('.link');

        expect(linkNode.text()).toBe('Закрыть');
    });

    it('should have default expanded label', () => {
        const collapse = mount(<Collapse isExpanded={ false }>Collapsing text</Collapse>);
        const linkNode = collapse.find('.link');

        expect(linkNode.text()).toBe('Expand');
    });

    it('should have default collapsed label', () => {
        const collapse = mount(<Collapse isExpanded={ true }>Collapsing text</Collapse>);
        const linkNode = collapse.find('.link');

        expect(linkNode.text()).toBe('Collapse');
    });
});
