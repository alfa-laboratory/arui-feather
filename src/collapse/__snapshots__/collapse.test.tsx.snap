// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapse should render without problem 1`] = `
<div
  className="collapse"
>
  <div
    className="collapse__content"
  >
    <div>
      Collapsing text
    </div>
    <ResizeSensor
      onResize={[Function]}
    />
  </div>
  <ForwardRef(ThemedComponent(Link))
    className="collapse__link"
    icon={
      <ForwardRef(ThemedComponent(IconArrowDown))
        size="s"
      />
    }
    iconPosition="right"
    onClick={[Function]}
    pseudo={true}
    text="Expand"
  />
</div>
`;
